### PR TITLE
Removed double close of one channel

### PR DESCRIPTION
**Pull request by david-littlefarmer**
Fix for this panic error:

panic: close of closed channel

goroutine 78 [running]:
github.com/goware/pubsub/googlebus.(*GoogleBus).Subscribe.func2()
	/go/pkg/mod/github.com/goware/pubsub@v0.7.0/googlebus/googlebus.go:221 +0x22b

### CODE REVIEW

**Comment by pkieltyka** [comment]
this stuff is super delicate and wonky, lets hold off

**Review by VojtechVitek — COMMENTED** [review]
LGTM.

Looks to be fixing the panic report.

Perhaps we can test this against real GoogleBus pub/sub and make sure it works as expected.

**Comment by pkieltyka** [comment]
the membus implements the same interface and close pattern https://github.com/goware/pubsub/tree/master/membus

it should be possible to reproduce the panic as a test case with the current code, then we can think of other edge cases and ensure our solution fixes it.

the problem here is a concurrency issue in order of closing.

**Comment by pkieltyka** [comment]
https://github.com/goware/pubsub/blob/master/membus/membus.go#L121

**Comment by david-littlefarmer** [comment]
I tried to reproduce the issue in membus. I added automatic cancel of context in membus after 4 seconds, just for demonstration purposes.
I pushed to this PR the TEMP code in membus, i also switch this PR to draft, until the temp code would be removed.

### Here is logs with same logic in membus as in googlesub, i get the panic:
```go
select {
case <-sub.done:
default:
	fmt.Println("closing sub.done")
	close(sub.done)
}

fmt.Println("unsub - canceling: ", sub.ChannelID())
cancel()
fmt.Println("unsub - canceled")
```
image

### Here is logs with same logic in membus as in googlesub with with my modification and i don't get a panic:
```go
fmt.Println("unsub - canceling: ", sub.ChannelID())
cancel()
fmt.Println("unsub - canceled")

<-sub.done
fmt.Println("unsub - sub.done filled")
```
image

With this pattern of gouroutine:
```go
go func() {
	defer close(sub.done)
	fmt.Println("subgo - defered close sub.done")

	// Receive messages from the Google PubSub subscription
	tempRunUntilCanceled(sctx)

	// In case of error, report it on the subscription and log
	fmt.Println("subgo - ending go routine")
}()
```

When log `subgo - ending go routine` occurs, then the `defer close(sub.done)` is triggered and then the panic happen. Because the closing of channel already happened in unsubscribe function.

**Comment by VojtechVitek** [comment]
@david-littlefarmer So you reproduced the panic in membus and proved the fix is correct, right?

**Comment by david-littlefarmer** [comment]
Yes.

File: membus/membus.go [hunk 1]
```diff
@@ -144,6 +170,28 @@ func (m *MemBus[M]) Subscribe(ctx context.Context, channelID string, optSubcript
...
+
+func tempRunUntilCanceled(ctx context.Context) {
+	for {
+		fmt.Println("worker - pubsubing")
+		time.Sleep(time.Second)
+
+		select {
+		case <-ctx.Done():
+			fmt.Println("worker - ctx canceled, return in 2s ")
+			time.Sleep(time.Second * 2)
```
Comment (pkieltyka) [inline membus/membus.go:188]: hmm.. so we are polling if the context is cancelled..?

**Comment by david-littlefarmer** [comment]
Closing, new solution will be in new PR.
